pytest/lib/peer.py

```diff
@@ -43,9 +43,12 @@ async def recv(self, expected=None):
             # Connection was closed on the other side
             if response_raw is None:
                 return None
-
-            response = BinarySerializer(schema).deserialize(
-                response_raw, PeerMessage)
+            try:
+                response = BinarySerializer(schema).deserialize(
+                    response_raw, PeerMessage)
+            except IndexError:
+                # unparsable message, ignore.
+                continue
```

Comment:
This feels wrong. You’re forcing the tests to use Borsh which means that you’re effectively disabling all those tests for protobufs plus this only postpones the problem since once Borsh gets disabled those tests will start failing again.

Reply:
I'm not forcing anything, these tests were using borsh and they are allowed to. These tests are poorly written, because they assume the underlying encoding by partially reimplementing the borsh encoding - and that's what "feels wrong". I'm not disabling anything just not providing extra coverage. I provide coverage from rust tests, python tests have only disadvantages.

Reply:
The tests were assuming Borsh because that was the network encoding node was using.  You’re now adding a new encoding without updating the tests to support it.  Since protobufs are going to be what nodes will use to talk with each other, you are effectively disabling testing of what is actually running in production.

Reply:
there is a difference between increasing amount of coverable logic and decreasing absolute coverage.

Reply:
Are you going to fix the tests once you remove Borsh support then? Why not do it now?

Reply:
> Are you going to fix the tests once you remove Borsh support then?

Definitely not leaving them broken, but I hoped to provide coverage with rust tests to get rid of this ad-hoc borsh implementation.

> Why not do it now?

Great question
1. Because it's weekend :)
2. Because I assumed you like nayduck green, so I wanted to fix the tests ASAP (without disabling them, because THAT would decrease the coverage)
3. Because this fix proves that borsh fallback works (and that is important).
4. Because I hate python and post-submit tests.

Reply:
OK, so can you file an Issue describing the tests need to be updated to use protobufs and add a TODO here?

By the way, with protobufs the code should be simpler since you can compile proto to Python and use that directly instead of having a hand-written definitions and deserialisers.

> Because this fix proves that borsh fallback works (and that is important).

There is upgradable.py which tests new and old nodes can communicate so that’s not as crucial.

> Because I hate python and post-submit tests.

You’re welcome to rewrite things in Rust.

Reply:
added TODO.

> By the way, with protobufs the code should be simpler since you can compile proto to Python and use that directly instead of having a hand-written definitions and deserialisers.

Yes, partially, the fix [WIP] is here https://github.com/near/nearcore/commits/gprusak-python-protobuf
The drawbacks are:
- I'm not removing borsh completely from communication, so non-network-related part of encoding is still there.
- I have no idea how to automate code generation on proto update, or even just enforce regenerating the python proto code on presubmit.